src/exporter/ceph_exporter.cc

```diff
@@ -64,8 +85,31 @@ int main(int argc, char **argv) {
   }
   common_init_finish(g_ceph_context);
 
+  // Start the web server thread
   boost::thread server_thread(web_server_thread_entrypoint);
-  DaemonMetricCollector &collector = collector_instance();
-  collector.main();
+
+  // Start the DaemonMetricCollector thread
+  boost::thread collector_thread(collector_thread_entrypoint);
+
+  // Main loop that waits for stop signals
+  while (!stop_signal_received.load()) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(500));
```

Comment:
Related to the comment on interrupting the execution of the timer callback instead of cancelling the timer, I'm not sure I see a reason for moving the metric collector into a separate thread and then having the main thread just uselessly wake up twice a second.  If the signal handler is changed to cancel `DaemonMetricCollector`'s timer and stop web server's `io_context`, it seems like the main thread can stay as it is today (i.e. run the metric collector).